File: recursion/javascript/solutions/zipLists.js
/**
 * Solution for zipLists.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function zipLists(list1, list2) {
  if (isEmpty(list1) || isEmpty(list2)) {
    return EMPTY_LIST;
  }

  let [first1, rest1] = unprepend(list1);
  let [first2, rest2] = unprepend(list2);

  return prepend([first1, first2], zipLists(rest1, rest2));
}

if (require.main === module) {
  let result = toArray(zipLists(Node.fromArray([1, 2, 3]), Node.fromArray(['a', 'b'])));
  console.assert(JSON.stringify(result) === '[[1,"a"],[2,"b"]]');

  console.assert(JSON.stringify(toArray(zipLists(EMPTY_LIST, Node.fromArray([1, 2])))) === '[]');
  console.assert(JSON.stringify(toArray(zipLists(Node.fromArray([1, 2]), EMPTY_LIST))) === '[]');

  console.log('All tests passed!');
}

module.exports = { zipLists };

File: recursion/python/solutions/zip_lists.py
"""Solution for zip_lists."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def zip_lists(lst1, lst2):
    """
    Combine two lists into list of tuples, stopping at shorter list.

    zip_lists(1 -> 2 -> 3, a -> b) => (1,a) -> (2,b)
    zip_lists((), 1 -> 2) => ()
    """
    if is_empty(lst1) or is_empty(lst2):
        return EMPTY_LIST

    first1, rest1 = unprepend(lst1)
    first2, rest2 = unprepend(lst2)

    return prepend((first1, first2), zip_lists(rest1, rest2))


if __name__ == '__main__':
    # Test cases
    assert to_python_list(zip_lists(Node.from_list([1, 2, 3]), Node.from_list(['a', 'b']))) == [(1, 'a'), (2, 'b')]
    assert to_python_list(zip_lists(EMPTY_LIST, Node.from_list([1, 2]))) == []
    assert to_python_list(zip_lists(Node.from_list([1, 2]), EMPTY_LIST)) == []
    assert to_python_list(zip_lists(Node.from_list([1, 2]), Node.from_list([3, 4]))) == [(1, 3), (2, 4)]

    print("All tests passed!")
